chore(async): Refactor EventEmitter and Async Facade

robwormald opened the pull request:
Refactor EventEmitter and Async Facade to match ES7 Observable semantics, properly use RxJS typedefs, make EventEmitter inherit from RxJS Subject. Closes #4149.

BREAKING CHANGE:
- consumers of EventEmitter no longer need to call .toRx()
- EventEmitter is now generic and requires a type - e.g. `EventEmitter<string>`
- EventEmitter and Observable now use the `.subscribe(generatorOrNext, error, complete)` method instead of `.observer(generator)`
- ObservableWrapper uses `callNext/callError/callComplete` instead of `callNext/callThrow/callReturn`

Notes: 
- This uses the RxJS main import path (though not the "kitchen sink"), and as such adds a number of methods to the public API (see public_api_spec). Note that Subject and EventEmitter both inherit from Observable and so it's less drastic than it looks in terms of new operators ;-) This is the current best workaround / compromise between having accurate type information and a usable EventEmitter / Observable interface. Reference discussions in https://github.com/angular/angular/issues/4390, and over in Rx https://github.com/ReactiveX/RxJS/issues/468 for a potential middle-term solution. This does add bytes to bundles, though with the incoming changes to bundling this could mostly be mitigated. In my opinion, this gives a better developer experience and we can watch over the beta period how and which operators should define the core. 
- Currently this fails on CI due to the dgeni generation of types not liking Rx's external .d.ts files - see https://github.com/angular/angular/issues/4967 but passes otherwise, and I'll rebase once that clears.
- It adds a psuedo Subject to the dart async facade that matches the EventEmitter - this could potentially be resolved by changes in ts2dart @mprobst ?
- tldr: this fixes a number of annoying issues re: developer experience (types and operators with EventEmitter, removes .toRx(), etc) and I think is a reasonable step to land before beta. I have a number of minor fixes (Http typedefs, etc) to follow to further improve said experience.

robwormald committed:
refactor(async): refactor EventEmitter

Refactor EventEmitter and Async Facade to match ES7 Observable semantics, properly use RxJS typedefs, make EventEmitter inherit from RxJS Subject. Closes #4149.

BREAKING CHANGE:
- consumers of EventEmitter no longer need to call .toRx()
- EventEmitter is now generic and requires a type - e.g. `EventEmitter<string>`
- EventEmitter and Observable now use the `.subscribe(generatorOrNext, error, complete)` method instead of `.observer(generator)`
- ObservableWrapper uses `callNext/callError/callComplete` instead of `callNext/callThrow/callReturn`

robwormald committed:
chore(package): update RxJS version

robwormald commented:
cc @jeffbcross per our discussions
cc @blesh per the "shipping Observable without operators is like shipping a keg of beer without a tap" commentary
cc @vsavkin to approve public API changeset.

naomiblack commented:
@vsavkin -- this previously had an LGTM from @mhevery but needs another review pass. Can you have a look?

mprobst commented on modules/angular2/src/core/facade/async.ts:
```diff
@@ -105,19 +98,32 @@ export class EventEmitter extends Observable {
+  }
 }
+
+export {Subject}
```
Nit: maybe move closer to the import?

mprobst commented:
I think the specific `Subject<T>` class in `async.dart` is a perfectly fine solution for this (AFAICT). Unless it causes any issues, I'd just go with this.

blesh commented:
Dart just needs an Rx library of it's own, IMO. You could even transpile the TypeScript from RxJS 5 over directly.

vicb commented:
> Dart just needs an Rx library of it's own, IMO. You could even transpile the TypeScript from RxJS 5 over directly.

May be https://github.com/frankpepermans/rxdart can help short term ?

blesh commented:
@vicb that library looks like it's wrapping an older version of RxJS and exposing it in Dart.

From my conversations with @IgorMinar, I *think* that the Angular team has the means to transpile TypeScript to Dart directly. Given that RxJS 5 is written entirely in TypeScript, and really isn't using any exotic JavaScript types, I think that it could be transpiled directly to Dart as well.

robwormald committed:
refactor(async): fix ObservableWrapper.isObservable

Makes ObservableWrapper and AsyncPipe work with Observable, Subject, and EventEmitter

mhevery commented:
@robwormald, can we get this green and merged.

robwormald committed:
chore(build): add RxJS to typings spec
